url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/Sender_multicast.java

```diff
@@ -0,0 +1,360 @@
+	File file = new File(fileName);
+	int lengthOfTheFile = (int) file.length();
+	
+	byte[] data = null;
+	try {
+	    data = Files.readAllBytes(file.toPath());
+	} catch (IOException e1) {
+	    System.err.println("Could not open file.");
+	    e1.printStackTrace();
+	    System.exit(1);
+	}
+	
+	// check loss
+	float percentageLoss = 99;
+	
+	// check overhead
+	int overhead = -1;
+	
+	try {
+	    overhead = Integer.valueOf(args[1]);
+	} catch (NumberFormatException e) {
+	    System.err.println("Invalid overhead value. (must be a positive integer)");
+	    System.exit(-1);
+	}
+	
+	if ( overhead < 0 ) {
+	    System.err.println("Invalid overhead value. (must be a positive integer)");
+	    System.exit(-1);
+	}
+	
+	// get IP and transform to InetAddress
+	InetAddress destIP = null;
+	
+	try {
+	    destIP = InetAddress.getByName(args[2]);
+	} catch (UnknownHostException e2) {
+	    e2.printStackTrace();
+	    System.err.println("invalid IP");
+	    System.exit(1);
+	}
+	int nbOfRecv;
+	
+	nbOfRecv = Integer.valueOf(args[4]);
+	if ( nbOfRecv < 0 ) {
+	    System.err.println("Invalid number of user (must be above 0)");
+	    System.exit(-1);
+	}
+	
+	// check destination port
+	int destPort = -1;
+	
+	try {
+	    destPort = Integer.valueOf(args[3]);
+	} catch (NumberFormatException e) {
+	    System.err.println("Invalid destination port. (must be above 1024)");
+	    System.exit(-1);
+	}
+	
+	if ( destPort < 1024 || destPort >= 65535 ) {
+	    System.err.println("Invalid destination port. (must be above 1024)");
+	    System.exit(-1);
+	}
+	
+	RecepteurThreadMulti receptionThread = new RecepteurThreadMulti(destPort + 1, destIP, nbOfRecv, historique);
+	
+	/**
+	 * preparation for the encoding
+	 */
+	
+	// create a new Encoder instance (usually one per file)
+	Encoder encoder = new Encoder(data, percentageLoss, overhead);
+	
+	// array that will contain the encoded symbols
+	EncodingPacket[] encodedSymbols = null;
+	
+	// array of source blocks
+	SourceBlock[] sourceBlocks = null;
+	int noBlocks;
+	
+	// total number of source symbols (for all source blocks)
+	int Kt = encoder.getKt();
+	
+	// partition the data into source blocks
+	sourceBlocks = encoder.partition();
+	noBlocks = sourceBlocks.length;
+	int nbPacket = -1;
+	// start the thread
+	receptionThread.start();
+	// long temps = System.currentTimeMillis() - before_2;
+	
+	// int normal_redundancy = (int) Math.round((((float) lengthOfTheFile / Utils.SYMB_LENGTH) * redondance) - ((float) lengthOfTheFile /
+	// Utils.SYMB_LENGTH));
+	boolean oneTime = true;
+	if ( oneTime ) {
+	    receptionThread.setTotalNumberOfPackets((int) Math.round((float) lengthOfTheFile / Utils.SYMB_LENGTH) + 1);
+	    oneTime = false;
+	}
+	
+	long before = System.currentTimeMillis();
+	while (receptionThread.getStatusEndLoop()) {
+	    
+	    /**
+	     * built of the socket
+	     */
+	    
+	    // open UDP socket
+	    MulticastSocket clientSocket = null;
+	    try {
+		clientSocket = new MulticastSocket();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *clientSocket*. The resource is referenced at lines 301, 318, 325. The resource is closed at line 340. There are other execution paths that do not contain closure statements, for example, when DatagramSocket.send throws an exception, and an else condition at lines 222, 269 evaluate to true. Ensure *clientSocket* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).